chapel-lang/chapel
# IFC: initial interface-constrained variant of PCGRandomStream

IFC: initial interface-constrained variant of PCGRandomStream

This introduces an interface `RandomStream`, implemented by PCGRandomStream,
and an interface-constrained variant of PCGRandomStream.fillRandom,
and also advances the implementation of interface generics
(mostly `interfaceResolution.cpp`) to support this code.

Test changes:

* adds `test/constrained-generics/random/`

  - `MyRandom.chpl` : a slight variation on the standard `Random.chpl`
  - `test-random.chpl` : the newly introduced interfaces, their implementations, and testing
  - `demo-random.chpl` : a slimmed-down version of test-random.chpl

* adjusts several .bad files to reflect the new behavior

While there:

* remove the commented out `INT_FATAL("CG case")` temporarily introduced in #17371

Future work:

* switch `holder` blocks to BLOCK_SCOPELESS, which should prevent them
  from carrying instantiation points and allow their complete removal

* factor out the testing of `icFillRandom()` into an IC function
  along the lines of `testGetNth()`, i.e., passing an array and a stream
  into IC-generic formals

Some development history is available at c95d6a9e20..0eb886f554
